netscalertool/netscalerapi.py

```diff
@@ -25,108 +25,87 @@ def __init__(self, args):
         for k, v in vars(args).iteritems():
             setattr(self, k, v)
 
-    def login(self):
+
+    def _call(self, method, url, payload=None, error_message=""):
         """
-        For debug purposes, print out the headers and the content of the
-        response
+        Dedupes much of the code necessary to facilitate api requests.
         """
 
-        # set the headers and the base URL
         headers = {'Content-type': 'application/json'}
-        url = "https://%s/nitro/v1/config/login" % self.host
+        if 'session_id' in dir(self):
+           headers['Cookie'] = 'sessionid='+self.session_id
 
-        # construct the payload with URL encoding
-        payload = {"login": {"username": self.user, "password":
-                   self.passwd}}
-        payload_encoded = json.dumps(payload)
+        payload_encoded = json.dumps(payload) if payload else None
 
-        # create a HTTP object, and use it to submit a POST request
         http = httplib2.Http(disable_ssl_certificate_validation=True)
         try:
-            response, content = http.request(url, 'POST', body=payload_encoded,
+            response, content = http.request(url, method, body=payload_encoded,
                                              headers=headers)
         except socket.error, e:
             msg = "Problem connecting to NetScaler %s:\n%s" % (self.host, e)
             raise RuntimeError(msg)
 
         data = json.loads(content)
-        errorcode = data["errorcode"]
+        errorcode = 0
+
+        try:
+            errorcode = data["errorcode"]
+        except:
+            print "Errorcode not found in response!"
 
         if response.status not in [200, 201] or errorcode != 0:
-            raise RuntimeError(content)
+            raise RuntimeError("%s: %s" % (error_message, content))
 
-        if self.debug:
-            print "Nitro API URL: ", url
-            print "\n", json.dumps(response)
-            print "\n", json.dumps(content)
+        return data
+
+
+    def login(self):
+        """
+        Starts a session with the netscaler.
+        """
+
+        url = "https://%s/nitro/v1/config/login" % self.host
+
+        payload = {"login": {"username": self.user, "password":
+                   self.passwd}}
+
+        data = self._call('POST', url, payload,
+                          error_message="\nCouldn't login")
 
-        data = json.loads(content)
         self.session_id = data["sessionid"]
 
+
     def logout(self):
         """
         Logout of the netscaler
         """
 
-        headers = {'Content-type': 'application/x-www-form-urlencoded',
-                   'Cookie': 'sessionid='+self.session_id}
-        url = "https://%s/nitro/v1/config/" % self.host
+        url = "https://%s/nitro/v1/config/logout" % self.host
```

Comment:
I suggest moving the first part of this into `_call`, since the `http://foo/nitro/v1' is going to be the same every time. Then `_call` would have a `path` parameter rather than `url`.

Reply:
That's fair.

Reply:
Instead of having the first half of the path (foo/nitro/v1) with the underling call method, I prefer letting the wrapper function manage that - I expect path to be the full path.